mirgecom/fluid.py

```diff
@@ -165,6 +166,21 @@ def join_conserved(dim, mass, energy, momentum,
     return result
 
 
+def join_conserved_vectors(dim, ary):
+    r"""Create a 2D array of shape(len(ary), dim)."""
+    neq = len(ary)
+    nspecies = neq - (dim+2)
+    retval = np.empty(shape=(neq, dim), dtype=object)
+    cv = split_conserved(dim, ary)
+    retval[0] = cv.mass
+    retval[1] = cv.energy
+    for i in range(dim):
+        retval[2+i] = ary[2+i]
+    for i in range(nspecies):
+        retval[dim+2+i] = ary[dim+2+i]
+    return retval
+
+
```

Comment:
Is there a way to write `join_conserved` to handle both of these cases?

Reply:
:shrug: Not that I could easily figure out... thus this beauty.

Reply:
I also would like to try to clean this up, but I do not understand this. Could you explain what the purpose of this [`_aux_shape` function is](https://github.com/illinois-ceesd/mirgecom/blob/main/mirgecom/fluid.py#L108-L122)? I'm trying to understand why [this is happening](https://github.com/illinois-ceesd/mirgecom/blob/main/mirgecom/fluid.py#L150-L159)

Reply:
I tried to see if `join_conserved` could be modified as @majosm suggested, and I was unable to get something working without resulting to extremely ugly hacks. However, I think _maybe_ it is worth using identical function signatures between `join_conserved` and `join_conserved_vectors`. This requires calling `split_conserved` before passing to `join_conserved_vectors`, but it's consistent with existing behavior.

I sketched out what I mean in this commit: https://github.com/illinois-ceesd/mirgecom/commit/1518e6dec0baa8b4dd4fa53262fb0b425ab4f1dc

Just a suggestion :)

Reply:
> I also would like to try to clean this up, but I do not understand this. First, I'm trying to wrap my head around `join_conserved` to see if this can be combined somehow. Could you explain what the purpose of this [`_aux_shape` function is](https://github.com/illinois-ceesd/mirgecom/blob/main/mirgecom/fluid.py?rgh-link-date=2021-04-27T23%3A14%3A59Z#L108-L122)? I'm trying to understand why [this is happening](https://github.com/illinois-ceesd/mirgecom/blob/main/mirgecom/fluid.py?rgh-link-date=2021-04-27T23%3A14%3A59Z#L150-L159)

Just a reminder about our slack convo:
https://ceesd.slack.com/archives/D01G09PTHN3/p1619574863021000

Reply:
> I tried to see if `join_conserved` could be modified as @majosm suggested, and I was unable to get something working without resorting to extremely ugly hacks. There probably exists a simple Pythonic way to maybe unify them, but at the cost of readability...? 🤷
> 
> However, I think _maybe_ it is worth using identical function signatures between `join_conserved` and `join_conserved_vectors`. This requires calling `split_conserved` before passing to `join_conserved_vectors`, but it's consistent with existing behavior.
> 
> I sketched out what I mean in this commit: [1518e6d](https://github.com/illinois-ceesd/mirgecom/commit/1518e6dec0baa8b4dd4fa53262fb0b425ab4f1dc)
> 
> Just a suggestion :)

This would be terrific, if it worked.  

The trouble is that when one goes:
```python
# for fluid q = [ mass, energy, mom1, mom2, mom3, spm1, spm2, spm3, spm4, .....] 
r_vol = obj_array_vectorize(weak_local_grad, q)
r_cv = split_conserved(dim, r_vol)
```
The arrays you get from `r_cv.momentum`, and `r_cv.species_mass` are not correct. You can see that I pick it apart manually, even after a call to `split_conserved` inside `join_conserved_vectors`.   The gradients of the scalars mass and energy come out arranged just fine, but the gradients of the vector components do not.

Reply:
[This](https://github.com/majosm/mirgecom/commit/d053e63e237f4b5badd7d58ddb07cc7c23c74277) seemed to work in my limited testing, but we should probably take a closer look to make sure.

Reply:
Works with single gas and mixtures.

Reply:
[Check this out](https://github.com/illinois-ceesd/mirgecom/pull/298/commits/c1c0024f74cf6fa2afabbe186f2175b5b452dc15), and w-hagen/mirgecom#16.

Reply:
Hooray! I like this. I think we found a winner